docs/developer-docs/smart-contracts/advanced-features/async-code.mdx

```diff
@@ -1,90 +1,94 @@
+
+* Always prefer best-effort response calls for calls that don't change the state of the callee, i.e., reads.
+* For endpoints that change the state of the callee, the best practice is to make such endpoints amenable to [safe retries](/docs/current/developer-docs/smart-contracts/best-practices/idempotency), since this is also required for safely handling external user calls. Use best-effort responses calls for such endpoints, and handle the additional edge cases.
+* Use guaranteed response calls for endpoints that mutate state and do not enable safe retries, or calls that perform larger cycle transfers. Best-effort response calls can currently lose cycles in edge cases. Be aware of the limitations of guaranteed response calls listed above. If safe upgrades are needed, consider using a stateless proxy canister (see the [security best practices](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/inter-canister-calls) for more information).
+
+## `async/await` syntax, concurrency and state changes
+
+The ``async/await`` syntax allows canister methods to issue asynchronous inter-canister calls and still be structured like an ordinary synchronous function. In this syntax, a method `foo` on a canister `A` could be written something like (in Rust-like syntax):
 
 ```
-/// Canister A
 async fn foo() {
-  baz().await;
+  do_work();
   call(canister_b, 'bar').await;
   do_more_work(res);
 }
+```
 
-async fn baz() {
-  do_local_work();
-}
+While this looks like an ordinary function, there are important differences in behavior between synchronous functions and functions that perform inter-canister calls.
 
-/// Canister B
-fn bar() {
-  do_some_more_work();
-}
-```
+As already mentioned, unlike, say, Ethereum, multiple method call executions on the same canister can be executed concurrently in the presence of inter-canister calls. This increases the canister's throughput, but it also means that the code needs to be correct also in the presence of concurrent behaviors. In particular, the developers have to ensure that no re-entrancy issues occur.
+
+Furthermore, canister methods that use inter-canister calls are **not atomic**. A failure somewhere in the method might not roll back all the changes that the method performed. Under the hood, such methods use the (low-level Wasm API)[### Low-level API] calls, and are translated into multiple *message handler Wasm functions*: an initial handler function that handles the method call itself (corresponding to `do_work` above), and other handler functions that serve as callbacks to handle the responses for any inter-canister calls that have been issued (`do_more_work` in the example above). The execution of each message handler is atomic on its own; an error (more precisely, a Wasm trap) rolls back only the changes performed by the current message handler, but does not affect the changes made by the other handlers.
```

Comment:
`(low-level Wasm API)[### Low-level API]` this link is broken in the rendering

Reply:
Fixed